File: test/location_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class LocationTest < Minitest::Spec
  test ".zero returns a fresh zero-valued location" do
    location = Herb::Location.zero

    assert_equal(
      { start: { line: 0, column: 0 }, end: { line: 0, column: 0 } },
      location.to_hash
    )

    refute_same location.start, location.end
    refute_same location, Herb::Location.zero
  end

  describe "#contains?" do
    test "answers for a position inside it" do
      assert Herb::Location[1, 0, 1, 5].contains?(Herb::Position[1, 3])
    end

    test "answers for its own start" do
      assert Herb::Location[1, 0, 1, 5].contains?(Herb::Position[1, 0])
    end

    test "does not answer for its own end" do
      refute Herb::Location[1, 0, 1, 5].contains?(Herb::Position[1, 5])
    end

    test "does not answer for a position before it" do
      refute Herb::Location[1, 2, 1, 5].contains?(Herb::Position[1, 1])
    end

    test "spans lines" do
      location = Herb::Location[1, 4, 3, 2]

      assert location.contains?(Herb::Position[2, 0])
      assert location.contains?(Herb::Position[3, 1])
      refute location.contains?(Herb::Position[3, 2])
      refute location.contains?(Herb::Position[1, 3])
    end
  end

  describe "#covers?" do
    test "answers for a location inside it" do
      assert Herb::Location[1, 0, 3, 0].covers?(Herb::Location[2, 0, 2, 4])
    end

    test "answers for itself" do
      assert Herb::Location[1, 0, 3, 0].covers?(Herb::Location[1, 0, 3, 0])
    end

    test "does not answer for a location that reaches past it" do
      refute Herb::Location[1, 0, 3, 0].covers?(Herb::Location[2, 0, 4, 0])
    end
  end

  describe "#empty?" do
    test "answers for a zero location" do
      assert_predicate Herb::Location.zero, :empty?
    end

    test "answers for any location that starts where it ends" do
      assert_predicate Herb::Location[2, 7, 2, 7], :empty?
    end

    test "does not answer for a location that spans a character" do
      refute_predicate Herb::Location[2, 7, 2, 8], :empty?
    end
  end
end

File: javascript/packages/core/test/location.test.ts
import { describe, test, expect } from "vitest"

import { Location } from "../src/location.js"
import { Position } from "../src/position.js"

describe("Location", () => {
  describe("contains", () => {
    test("answers for a position inside it", () => {
      expect(Location.from(1, 0, 1, 5).contains(Position.from(1, 3))).toBe(true)
    })

    test("answers for its own start", () => {
      expect(Location.from(1, 0, 1, 5).contains(Position.from(1, 0))).toBe(true)
    })

    test("does not answer for its own end", () => {
      expect(Location.from(1, 0, 1, 5).contains(Position.from(1, 5))).toBe(false)
    })

    test("does not answer for a position before it", () => {
      expect(Location.from(1, 2, 1, 5).contains(Position.from(1, 1))).toBe(false)
    })

    test("spans lines", () => {
      const location = Location.from(1, 4, 3, 2)

      expect(location.contains(Position.from(2, 0))).toBe(true)
      expect(location.contains(Position.from(3, 1))).toBe(true)
      expect(location.contains(Position.from(3, 2))).toBe(false)
      expect(location.contains(Position.from(1, 3))).toBe(false)
    })
  })

  describe("covers", () => {
    test("answers for a location inside it", () => {
      expect(Location.from(1, 0, 3, 0).covers(Location.from(2, 0, 2, 4))).toBe(true)
    })

    test("answers for itself", () => {
      expect(Location.from(1, 0, 3, 0).covers(Location.from(1, 0, 3, 0))).toBe(true)
    })

    test("does not answer for a location that reaches past it", () => {
      expect(Location.from(1, 0, 3, 0).covers(Location.from(2, 0, 4, 0))).toBe(false)
    })
  })

  describe("isEmpty", () => {
    test("answers for a zero location", () => {
      expect(Location.zero.isEmpty()).toBe(true)
    })

    test("answers for any location that starts where it ends", () => {
      expect(Location.from(2, 7, 2, 7).isEmpty()).toBe(true)
    })

    test("does not answer for a location that spans a character", () => {
      expect(Location.from(2, 7, 2, 8).isEmpty()).toBe(false)
    })
  })
})
